File: 03_NLRM_s_jinou_apriorni_hustotou_importance_sampling/cv05_pr01.py
# Nastaveni cesty do domovske slozky
import sys
root_dir = "/".join(sys.argv[0].split("/")[:-3])
sys.path.insert(1, root_dir + "/Support/Python")

# importy pomocnych funkci
from progress_info import progress_bar
from my_boot import my_boot

# importy oficialnich knihoven
import numpy as np
from numpy import transpose as t
from math import exp
from tabulate import tabulate
import matplotlib.pyplot as plt

# ===== 1. Posteriorni hustota N(0,1) + Importance function U(a,b) =====
# Nastaveni importance sampler
s = 10_000
w = np.zeros((1,s))         # radkovy vektor vah
theta = np.zeros((1,s))     # vektor vzorku z importance function

# Importance function U(a,b)
a = -100
b = 100

# ===== 2. Importance sampling =====
for i in range(s):
    # a) Generovani vzorku z importance function, q(theta)
    theta[0,i] = np.random.rand() * (b - a) + a
    # b) Vypocet vah (na zaklade jadrovych hustot)
    p_theta = exp(-1 / 2 * theta[0,i]**2)
    q_theta = 1 / (b - a)
    w[0,i] = p_theta / q_theta
    progress_bar(i,s)

# ===== 3. Vypocet posteriorni stredni hodnoty a rozptylu =====

e_theta = w @ t(theta) / np.sum(w)          # vazena stredni hodnota
e_theta_2 = w @ (t(theta)**2) / np.sum(w)   # vazena stredni hodnota "na druhou"
d_theta = e_theta_2 - e_theta**2            # vazeny rozptyl
w_avg = np.mean(w)                          # prumerna vaha

headers = ["E[theta]", "D[theta]", "E[w]", "Vel. vzorku"]
table = [[round(e_theta[0][0], 4), round(d_theta[0][0], 4), round(w_avg, 4), s]]    # Musi byt list listu kvuli vytvoreni tabulky
print(f"\nImportance sampling s vyuzitim U({a},{b})")
print(tabulate(table, headers, tablefmt="pretty"))


# ===== 4. Vazeny bootstrap pro ziskani vyberu z posteriorni hodnoty =====
boot_smpl = my_boot(theta, s, w)
bootstrapped_theta = theta[0][boot_smpl]

fig, axs = plt.subplots(2,1, tight_layout=True)

axs[0].hist(theta, bins=50, facecolor="blue", alpha=0.5, edgecolor="black", histtype="step")
axs[1].hist(bootstrapped_theta, bins=50, facecolor="blue", histtype="step")

plt.show()

File: Témata/03 NLRM s jinou apriorní hustotou (Importance sampling)/Matlab/cv05_pr01.m
%Generovani momentu a vzorku N(0,1) z U(a,b)
 clear
 close all
 clc

%% Nastaveni cest
 addpath('.\Support')

%% Posteriorni hustotu N(0,1) + importance function U(a,b)
%nastaveni importance sampler
 S = 100000;
 w = zeros(1,S); %radkovy vektor vah
 theta = zeros(1,S); %vektor vzorku z importance function
 
%importance function U(a,b)
 a = -100;
 b = 100;

%% Importance sampling
for s = 1:S
   %1. generovani vzorku z importance function, q(theta)
   theta(s) = rand*(b-a)+a;
   %2. vypocet vah (na zaklade jadrovych hustot)
   p_theta = exp(-1/2*theta(s)^2); %jadrova hustota
   q_theta = 1/(b-a); %plna funkce hustoty (staci i jadro = 1)
   w(s) = p_theta/q_theta;    
end

%% Vypocet posteriorni stredni hodnoty a rozptylu
 E_theta = w*theta'/sum(w);
 E_theta_2 = w*(theta'.^2)/sum(w); %vazena E(theta^2)
 D_theta = E_theta_2-E_theta^2; %vazeny rozptyl
 w_avg = mean(w); %prumerna vaha

 fprintf('Importance sampling s vyuzitim U(%5.0f,%5.0f)\n',a,b)
 fprintf('E_theta   D_theta    w_avg          S\n')
 fprintf('%6.4f    %6.4f     %6.4f  %10.0f\n',E_theta,D_theta,w_avg,S)

%% Vazeny bootstrap pro ziskani vyberu z posteriorni hodnoty
 boot_smpl = my_boot(theta,length(theta),w);

figure
 subplot(2,1,1)
 histogram(theta,50)
 title('Vyber z importance function')
 
 subplot(2,1,2)
 histogram(theta(boot_smpl),50)
 title('Vyber z bootstrapovaneho rozdeleni')
 
% Vypocet strednich hodnot a rozptylu z bootstrapovaneho rozdeleni
 E_boot_theta = mean(theta(boot_smpl));
 D_boot_theta = var(theta(boot_smpl));
 
 fprintf('Momenty z boostrapovaneho rozdeleni\n')
 fprintf('E_boot_theta   D_boot_theta\n')
 fprintf('%6.4f           %6.4f     \n',E_boot_theta,D_boot_theta)

 

